{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28f09c18-e8cf-48b3-bb5e-937b386c7280",
   "metadata": {},
   "source": [
    "# Figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a16c2ad-0c12-4120-a884-40f7cd476abe",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.stats as stats\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scvelo as scv\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap\n",
    "from matplotlib_venn import venn2\n",
    "from venny4py.venny4py import *\n",
    "from scipy.spatial.distance import cosine\n",
    "from scipy.stats import wasserstein_distance\n",
    "from scipy.spatial import distance\n",
    "from scipy.stats import pearsonr\n",
    "import matplotlib as mpl\n",
    "from scvelo.core import LinearRegression"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a500ef8-b3d0-4021-976a-c50566536da5",
   "metadata": {},
   "source": [
    "## load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4af87762-b79f-4c10-981e-e84559f86004",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stewart\n",
    "stewart_velocyto = sc.read_h5ad(\"../Stewart/data/stewart_velocyto.h5ad\")\n",
    "stewart_alevin = sc.read_h5ad(\"../Stewart/data/stewart_alevin.h5ad\")\n",
    "\n",
    "# Mathew\n",
    "mathew_velocyto = sc.read_h5ad(\"../Mathew/data/mathew_velocyto.h5ad\")\n",
    "mathew_alevin = sc.read_h5ad(\"../Mathew/data/mathew_alevin.h5ad\")\n",
    "\n",
    "# Fu\n",
    "fu_velocyto = sc.read_h5ad(\"../Fu/data/fu_velocyto.h5ad\")\n",
    "fu_alevin = sc.read_h5ad(\"../Fu/data/fu_alevin.h5ad\")\n",
    "\n",
    "# separate Fu into paired-end and single-end alignment\n",
    "fu_pe_velocyto = fu_velocyto[~fu_velocyto.obs[\"id\"].isin([\"MJ001\", \"MJ002\", \"MJ003\", \"MJ017\", \"MJ016\"]),:]\n",
    "fu_se_velocyto = fu_velocyto[fu_velocyto.obs[\"id\"].isin([\"MJ001\", \"MJ002\", \"MJ003\", \"MJ017\", \"MJ016\"]),:]\n",
    "fu_pe_alevin = fu_alevin[~fu_alevin.obs[\"id\"].isin([\"MJ001\", \"MJ002\", \"MJ003\", \"MJ017\", \"MJ016\"]),:]\n",
    "fu_se_alevin = fu_alevin[fu_alevin.obs[\"id\"].isin([\"MJ001\", \"MJ002\", \"MJ003\", \"MJ017\", \"MJ016\"]),:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f672e31-2d81-4e9a-9590-e95a8430b6fb",
   "metadata": {},
   "source": [
    "## Figure 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf0e40ae-5671-4723-b077-3a4a2c232313",
   "metadata": {},
   "source": [
    "### Figure 1a: total counts (u+s+a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f82002f-ef4d-406e-9841-5df8f21001ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get u,s,a counts per cell\n",
    "\n",
    "stewart_velocyto_spliced = np.array(stewart_velocyto.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "stewart_velocyto_unspliced = np.array(stewart_velocyto.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "stewart_velocyto_ambiguous = np.array(stewart_velocyto.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "stewart_alevin_spliced = np.array(stewart_alevin.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "stewart_alevin_unspliced = np.array(stewart_alevin.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "stewart_alevin_ambiguous = np.array(stewart_alevin.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "mathew_velocyto_spliced = np.array(mathew_velocyto.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "mathew_velocyto_unspliced = np.array(mathew_velocyto.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "mathew_velocyto_ambiguous = np.array(mathew_velocyto.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "mathew_alevin_spliced = np.array(mathew_alevin.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "mathew_alevin_unspliced = np.array(mathew_alevin.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "mathew_alevin_ambiguous = np.array(mathew_alevin.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_velocyto_spliced = np.array(fu_velocyto.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "fu_velocyto_unspliced = np.array(fu_velocyto.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "fu_velocyto_ambiguous = np.array(fu_velocyto.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_alevin_spliced = np.array(fu_alevin.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "fu_alevin_unspliced = np.array(fu_alevin.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "fu_alevin_ambiguous = np.array(fu_alevin.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_pe_velocyto_spliced = np.array(fu_pe_velocyto.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "fu_pe_velocyto_unspliced = np.array(fu_pe_velocyto.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "fu_pe_velocyto_ambiguous = np.array(fu_pe_velocyto.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_pe_alevin_spliced = np.array(fu_pe_alevin.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "fu_pe_alevin_unspliced = np.array(fu_pe_alevin.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "fu_pe_alevin_ambiguous = np.array(fu_pe_alevin.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_se_velocyto_spliced = np.array(fu_se_velocyto.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "fu_se_velocyto_unspliced = np.array(fu_se_velocyto.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "fu_se_velocyto_ambiguous = np.array(fu_se_velocyto.layers[\"ambiguous\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_se_alevin_spliced = np.array(fu_se_alevin.layers[\"spliced\"].sum(axis=1)).flatten()\n",
    "fu_se_alevin_unspliced = np.array(fu_se_alevin.layers[\"unspliced\"].sum(axis=1)).flatten()\n",
    "fu_se_alevin_ambiguous = np.array(fu_se_alevin.layers[\"ambiguous\"].sum(axis=1)).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24647f8-ccda-4327-83df-7291d961ca2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get total counts per cell\n",
    "stewart_velocyto_counts = stewart_velocyto_spliced + stewart_velocyto_unspliced + stewart_velocyto_ambiguous\n",
    "stewart_alevin_counts = stewart_alevin_spliced + stewart_alevin_unspliced + stewart_alevin_ambiguous\n",
    "\n",
    "mathew_velocyto_counts = mathew_velocyto_spliced + mathew_velocyto_unspliced + mathew_velocyto_ambiguous\n",
    "mathew_alevin_counts = mathew_alevin_spliced + mathew_alevin_unspliced + mathew_alevin_ambiguous\n",
    "\n",
    "fu_velocyto_counts = fu_velocyto_spliced + fu_velocyto_unspliced + fu_velocyto_ambiguous\n",
    "fu_alevin_counts = fu_alevin_spliced + fu_alevin_unspliced + fu_alevin_ambiguous\n",
    "\n",
    "fu_pe_velocyto_counts = fu_pe_velocyto_spliced + fu_pe_velocyto_unspliced + fu_pe_velocyto_ambiguous\n",
    "fu_pe_alevin_counts = fu_pe_alevin_spliced + fu_pe_alevin_unspliced + fu_pe_alevin_ambiguous\n",
    "\n",
    "fu_se_velocyto_counts = fu_se_velocyto_spliced + fu_se_velocyto_unspliced + fu_se_velocyto_ambiguous\n",
    "fu_se_alevin_counts = fu_se_alevin_spliced + fu_se_alevin_unspliced + fu_se_alevin_ambiguous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b64a3066-3d16-42b6-89f9-b071fd5655f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Total UMIs velocyto Stewart: \", stewart_velocyto_counts.sum())\n",
    "print(\"Total UMIs alevin   Stewart: \", stewart_alevin_counts.sum())\n",
    "print(\"Ratio Stewart: \", stewart_alevin_counts.sum() / stewart_velocyto_counts.sum(), \"\\n\")\n",
    "\n",
    "print(\"Total UMIs velocyto Mathew:  \", mathew_velocyto_counts.sum())\n",
    "print(\"Total UMIs alevin   Mathew:  \", mathew_alevin_counts.sum())\n",
    "print(\"Ratio Mathew: \", mathew_alevin_counts.sum() / mathew_velocyto_counts.sum(), \"\\n\")\n",
    "\n",
    "print(\"Total UMIs velocyto Fu (SE): \", fu_se_velocyto_counts.sum())\n",
    "print(\"Total UMIs alevin   Fu (SE): \", fu_se_alevin_counts.sum())\n",
    "print(\"Ratio Fu (SE): \", fu_se_alevin_counts.sum() / fu_se_velocyto_counts.sum(), \"\\n\")\n",
    "\n",
    "print(\"Total UMIs velocyto Fu (PE): \", fu_pe_velocyto_counts.sum())\n",
    "print(\"Total UMIs alevin   Fu (PE): \", fu_pe_alevin_counts.sum())\n",
    "print(\"Ratio Fu (PE): \", fu_pe_alevin_counts.sum() / fu_pe_velocyto_counts.sum(), \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d03fc26e-bae1-4de8-9b29-5b4831d7f7d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get total cellranger counts\n",
    "sc.pp.calculate_qc_metrics(stewart_velocyto, layer=\"counts\", inplace=True, log1p=False)\n",
    "sc.pp.calculate_qc_metrics(mathew_velocyto, layer=\"counts\", inplace=True, log1p=False)\n",
    "sc.pp.calculate_qc_metrics(fu_velocyto, layer=\"counts\", inplace=True, log1p=False)\n",
    "sc.pp.calculate_qc_metrics(fu_pe_velocyto, layer=\"counts\", inplace=True, log1p=False)\n",
    "sc.pp.calculate_qc_metrics(fu_se_velocyto, layer=\"counts\", inplace=True, log1p=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40797fcc-c777-42ce-b210-36eb7282ab7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for boxplot\n",
    "fig1a = pd.DataFrame({\"Counts\": np.concatenate([stewart_velocyto_counts, stewart_alevin_counts, stewart_velocyto.obs[\"total_counts\"].values, \n",
    "                                                mathew_velocyto_counts, mathew_alevin_counts, mathew_velocyto.obs[\"total_counts\"].values,\n",
    "                                                fu_se_velocyto_counts, fu_se_alevin_counts, fu_se_velocyto.obs[\"total_counts\"].values,\n",
    "                                                fu_pe_velocyto_counts, fu_pe_alevin_counts, fu_pe_velocyto.obs[\"total_counts\"].values]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_obs * 3,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_obs * 3,\n",
    "                                                 [\"Fu et al. (SE)\"] * fu_se_velocyto.n_obs * 3,\n",
    "                                                 [\"Fu et al. (PE)\"] * fu_pe_velocyto.n_obs * 3]),\n",
    "                      \"Method\": np.concatenate([[\"Velocyto\"] * stewart_velocyto.n_obs, [\"Alevin-fry\"] * stewart_velocyto.n_obs, [\"Cellranger\"] * stewart_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * mathew_velocyto.n_obs, [\"Alevin-fry\"] * mathew_velocyto.n_obs, [\"Cellranger\"] * mathew_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * fu_se_velocyto.n_obs, [\"Alevin-fry\"] * fu_se_velocyto.n_obs, [\"Cellranger\"] * fu_se_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * fu_pe_velocyto.n_obs, [\"Alevin-fry\"] * fu_pe_velocyto.n_obs, [\"Cellranger\"] * fu_pe_velocyto.n_obs])})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6033fe04-8ec6-4a21-96a5-d682116ec620",
   "metadata": {},
   "source": [
    "### Figure 1b: difference cellranger/salmon counts - u+s+a counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5442b76-3294-4adf-be60-711eec236a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get cellranger/salmon total counts per cell\n",
    "stewart_cellranger_total = np.array(stewart_velocyto.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "stewart_salmon_total = np.array(stewart_alevin.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "\n",
    "mathew_cellranger_total = np.array(mathew_velocyto.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "mathew_salmon_total = np.array(mathew_alevin.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_cellranger_total = np.array(fu_velocyto.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "fu_salmon_total = np.array(fu_alevin.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_se_cellranger_total = np.array(fu_se_velocyto.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "fu_se_salmon_total = np.array(fu_se_alevin.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "\n",
    "fu_pe_cellranger_total = np.array(fu_pe_velocyto.layers[\"counts\"].sum(axis=1)).flatten()\n",
    "fu_pe_salmon_total = np.array(fu_pe_alevin.layers[\"counts\"].sum(axis=1)).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7903373-d993-4183-a758-ba852bc80d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get difference per cell\n",
    "stewart_velocyto_diff = stewart_cellranger_total - stewart_velocyto_counts\n",
    "stewart_alevin_diff = stewart_salmon_total - stewart_alevin_counts\n",
    "\n",
    "mathew_velocyto_diff = mathew_cellranger_total - mathew_velocyto_counts\n",
    "mathew_alevin_diff = mathew_salmon_total - mathew_alevin_counts\n",
    "\n",
    "fu_velocyto_diff = fu_cellranger_total - fu_velocyto_counts\n",
    "fu_alevin_diff = fu_salmon_total - fu_alevin_counts\n",
    "\n",
    "fu_se_velocyto_diff = fu_se_cellranger_total - fu_se_velocyto_counts\n",
    "fu_se_alevin_diff = fu_se_salmon_total - fu_se_alevin_counts\n",
    "\n",
    "fu_pe_velocyto_diff = fu_pe_cellranger_total - fu_pe_velocyto_counts\n",
    "fu_pe_alevin_diff = fu_pe_salmon_total - fu_pe_alevin_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8dcd9c2-c1ed-4d4d-a002-9eee3188fdb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for violinplot\n",
    "fig1b = pd.DataFrame({\"Pct\": np.concatenate([stewart_velocyto_diff, stewart_alevin_diff,\n",
    "                                             mathew_velocyto_diff, mathew_alevin_diff,\n",
    "                                             fu_se_velocyto_diff, fu_se_alevin_diff,\n",
    "                                             fu_pe_velocyto_diff, fu_pe_alevin_diff]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_obs *2,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_obs *2,\n",
    "                                                 [\"Fu et al. (SE)\"] * fu_se_velocyto.n_obs * 2,\n",
    "                                                 [\"Fu et al. (PE)\"] * fu_pe_velocyto.n_obs * 2]),\n",
    "                      \"Method\": np.concatenate([[\"Velocyto\"] * stewart_velocyto.n_obs, [\"Alevin-fry\"] * stewart_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * mathew_velocyto.n_obs, [\"Alevin-fry\"] * mathew_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * fu_se_velocyto.n_obs, [\"Alevin-fry\"] * fu_se_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * fu_pe_velocyto.n_obs, [\"Alevin-fry\"] * fu_pe_velocyto.n_obs])})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46378320-3ac9-4095-b294-ddb4f01dfac8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get means\n",
    "print(\"Mean difference velocyto:\")\n",
    "print(\"Stewart et al.:\", fig1b.loc[(fig1b[\"Method\"] == \"Velocyto\") & (fig1b[\"Dataset\"] == \"Stewart et al.\"), \"Pct\"].mean())\n",
    "print(\"Mathew et al.:\", fig1b.loc[(fig1b[\"Method\"] == \"Velocyto\") & (fig1b[\"Dataset\"] == \"Mathew et al.\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (SE):\", fig1b.loc[(fig1b[\"Method\"] == \"Velocyto\") & (fig1b[\"Dataset\"] == \"Fu et al. (SE)\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (PE):\", fig1b.loc[(fig1b[\"Method\"] == \"Velocyto\") & (fig1b[\"Dataset\"] == \"Fu et al. (PE)\"), \"Pct\"].mean())\n",
    "\n",
    "print(\"Mean difference alevin:\")\n",
    "print(\"Stewart et al.:\", fig1b.loc[(fig1b[\"Method\"] == \"Alevin-fry\") & (fig1b[\"Dataset\"] == \"Stewart et al.\"), \"Pct\"].mean())\n",
    "print(\"Mathew et al.:\", fig1b.loc[(fig1b[\"Method\"] == \"Alevin-fry\") & (fig1b[\"Dataset\"] == \"Mathew et al.\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (SE):\", fig1b.loc[(fig1b[\"Method\"] == \"Alevin-fry\") & (fig1b[\"Dataset\"] == \"Fu et al. (SE)\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (PE):\", fig1b.loc[(fig1b[\"Method\"] == \"Alevin-fry\") & (fig1b[\"Dataset\"] == \"Fu et al. (PE)\"), \"Pct\"].mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ecae8d9a-e416-4b8e-b8c6-ea2f7a37076a",
   "metadata": {},
   "source": [
    "### Figure 1c: correlation of counts with spliced counts of same gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9f4047c-db5d-49ca-a990-41e2dd90ff85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get correlations\n",
    "\n",
    "stewart_velo_corr = [pearsonr(stewart_velocyto[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                             stewart_velocyto[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                     for gene in stewart_velocyto.var_names]\n",
    "\n",
    "stewart_alevin_corr = [pearsonr(stewart_alevin[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                                stewart_alevin[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                       for gene in stewart_alevin.var_names]\n",
    "\n",
    "mathew_velo_corr = [pearsonr(mathew_velocyto[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                             mathew_velocyto[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                    for gene in mathew_velocyto.var_names]\n",
    "\n",
    "mathew_alevin_corr = [pearsonr(mathew_alevin[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                               mathew_alevin[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                      for gene in mathew_alevin.var_names]\n",
    "\n",
    "fu_se_velo_corr = [pearsonr(fu_se_velocyto[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                            fu_se_velocyto[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                   for gene in fu_se_velocyto.var_names]\n",
    "\n",
    "fu_se_alevin_corr = [pearsonr(fu_se_alevin[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                              fu_se_alevin[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                     for gene in fu_se_alevin.var_names]\n",
    "\n",
    "fu_pe_velo_corr = [pearsonr(fu_pe_velocyto[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                            fu_pe_velocyto[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                   for gene in fu_pe_velocyto.var_names]\n",
    "\n",
    "fu_pe_alevin_corr = [pearsonr(fu_pe_alevin[:,gene].layers[\"counts\"].toarray().ravel(),\n",
    "                              fu_pe_alevin[:,gene].layers[\"spliced\"].toarray().ravel())[0] \n",
    "                     for gene in fu_pe_alevin.var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22cea0e4-1cd5-48cb-8573-d03aa943ca45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for boxplot\n",
    "fig1c = pd.DataFrame({\"Pearson correlation\": np.concatenate([stewart_velo_corr, stewart_alevin_corr,\n",
    "                                                             mathew_velo_corr, mathew_alevin_corr,\n",
    "                                                             fu_se_velo_corr, fu_se_alevin_corr,\n",
    "                                                             fu_pe_velo_corr, fu_pe_alevin_corr ]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_vars *2,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_vars *2,\n",
    "                                                 [\"Fu et al. (SE)\"] * fu_se_velocyto.n_vars * 2,\n",
    "                                                 [\"Fu et al. (PE)\"] * fu_pe_velocyto.n_vars * 2]),\n",
    "                      \"Method\": np.concatenate([[\"Velocyto\"] * stewart_velocyto.n_vars, [\"Alevin-fry\"] * stewart_velocyto.n_vars,\n",
    "                                                [\"Velocyto\"] * mathew_velocyto.n_vars, [\"Alevin-fry\"] * mathew_velocyto.n_vars,\n",
    "                                                [\"Velocyto\"] * fu_se_velocyto.n_vars, [\"Alevin-fry\"] * fu_se_velocyto.n_vars,\n",
    "                                                [\"Velocyto\"] * fu_pe_velocyto.n_vars, [\"Alevin-fry\"] * fu_pe_velocyto.n_vars])})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bacf911-c695-4f82-a261-34309c1ad7bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get means\n",
    "print(\"Mean correlation velocyto:\")\n",
    "print(\"Stewart et al.:\", fig1c.loc[(fig1c[\"Method\"] == \"Velocyto\") & (fig1c[\"Dataset\"] == \"Stewart et al.\"), \"Pearson correlation\"].mean())\n",
    "print(\"Mathew et al.:\", fig1c.loc[(fig1c[\"Method\"] == \"Velocyto\") & (fig1c[\"Dataset\"] == \"Mathew et al.\"), \"Pearson correlation\"].mean())\n",
    "print(\"Fu et al. (SE):\", fig1c.loc[(fig1c[\"Method\"] == \"Velocyto\") & (fig1c[\"Dataset\"] == \"Fu et al. (SE)\"), \"Pearson correlation\"].mean())\n",
    "print(\"Fu et al. (PE):\", fig1c.loc[(fig1c[\"Method\"] == \"Velocyto\") & (fig1c[\"Dataset\"] == \"Fu et al. (PE)\"), \"Pearson correlation\"].mean())\n",
    "\n",
    "print(\"Mean correlation alevin:\")\n",
    "print(\"Stewart et al.:\", fig1c.loc[(fig1c[\"Method\"] == \"Alevin-fry\") & (fig1c[\"Dataset\"] == \"Stewart et al.\"), \"Pearson correlation\"].mean())\n",
    "print(\"Mathew et al.:\", fig1c.loc[(fig1c[\"Method\"] == \"Alevin-fry\") & (fig1c[\"Dataset\"] == \"Mathew et al.\"), \"Pearson correlation\"].mean())\n",
    "print(\"Fu et al. (SE):\", fig1c.loc[(fig1c[\"Method\"] == \"Alevin-fry\") & (fig1c[\"Dataset\"] == \"Fu et al. (SE)\"), \"Pearson correlation\"].mean())\n",
    "print(\"Fu et al. (PE):\", fig1c.loc[(fig1c[\"Method\"] == \"Alevin-fry\") & (fig1c[\"Dataset\"] == \"Fu et al. (PE)\"), \"Pearson correlation\"].mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc86c82a-6a96-4e02-8f0b-6de910c293ad",
   "metadata": {},
   "source": [
    "### Figure 1d: ratio spliced/unspliced"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f60d0059-cdc0-440d-a02c-ec805c2fb975",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get ratios per cell\n",
    "stewart_velocyto_pct_unspliced = (stewart_velocyto_unspliced  / (stewart_velocyto_spliced + stewart_velocyto_unspliced)) * 100\n",
    "stewart_alevin_pct_unspliced = (stewart_alevin_unspliced  / (stewart_alevin_spliced + stewart_alevin_unspliced)) * 100\n",
    "\n",
    "mathew_velocyto_pct_unspliced = (mathew_velocyto_unspliced  / (mathew_velocyto_spliced + mathew_velocyto_unspliced)) * 100\n",
    "mathew_alevin_pct_unspliced = (mathew_alevin_unspliced  / (mathew_alevin_spliced + mathew_alevin_unspliced)) * 100\n",
    "\n",
    "fu_velocyto_pct_unspliced = (fu_velocyto_unspliced  / (fu_velocyto_spliced + fu_velocyto_unspliced)) * 100\n",
    "fu_alevin_pct_unspliced = (fu_alevin_unspliced  / (fu_alevin_spliced + fu_alevin_unspliced)) * 100\n",
    "\n",
    "fu_se_velocyto_pct_unspliced = (fu_se_velocyto_unspliced  / (fu_se_velocyto_spliced + fu_se_velocyto_unspliced)) * 100\n",
    "fu_se_alevin_pct_unspliced = (fu_se_alevin_unspliced  / (fu_se_alevin_spliced + fu_se_alevin_unspliced)) * 100\n",
    "\n",
    "fu_pe_velocyto_pct_unspliced = (fu_pe_velocyto_unspliced  / (fu_pe_velocyto_spliced + fu_pe_velocyto_unspliced)) * 100\n",
    "fu_pe_alevin_pct_unspliced = (fu_pe_alevin_unspliced  / (fu_pe_alevin_spliced + fu_pe_alevin_unspliced)) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d123f2a0-aa4a-4358-8378-dcfa2b0998d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for violinplot\n",
    "fig1d = pd.DataFrame({\"Pct\": np.concatenate([stewart_velocyto_pct_unspliced, stewart_alevin_pct_unspliced,\n",
    "                                             mathew_velocyto_pct_unspliced, mathew_alevin_pct_unspliced,\n",
    "                                             fu_se_velocyto_pct_unspliced, fu_se_alevin_pct_unspliced,\n",
    "                                             fu_pe_velocyto_pct_unspliced, fu_pe_alevin_pct_unspliced]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_obs *2,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_obs *2,\n",
    "                                                 [\"Fu et al. (SE)\"] * fu_se_velocyto.n_obs * 2,\n",
    "                                                 [\"Fu et al. (PE)\"] * fu_pe_velocyto.n_obs * 2]),\n",
    "                      \"Method\": np.concatenate([[\"Velocyto\"] * stewart_velocyto.n_obs, [\"Alevin-fry\"] * stewart_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * mathew_velocyto.n_obs, [\"Alevin-fry\"] * mathew_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * fu_se_velocyto.n_obs, [\"Alevin-fry\"] * fu_se_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * fu_pe_velocyto.n_obs, [\"Alevin-fry\"] * fu_pe_velocyto.n_obs])})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c214f03e-32ab-49b5-86f7-b97a34f6d155",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get means\n",
    "print(\"Mean unspliced ratio velocyto:\")\n",
    "print(\"Stewart et al.:\", fig1d.loc[(fig1d[\"Method\"] == \"Velocyto\") & (fig1d[\"Dataset\"] == \"Stewart et al.\"), \"Pct\"].mean())\n",
    "print(\"Mathew et al.:\", fig1d.loc[(fig1d[\"Method\"] == \"Velocyto\") & (fig1d[\"Dataset\"] == \"Mathew et al.\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (SE):\", fig1d.loc[(fig1d[\"Method\"] == \"Velocyto\") & (fig1d[\"Dataset\"] == \"Fu et al. (SE)\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (PE):\", fig1d.loc[(fig1d[\"Method\"] == \"Velocyto\") & (fig1d[\"Dataset\"] == \"Fu et al. (PE)\"), \"Pct\"].mean())\n",
    "\n",
    "print(\"Mean unspliced ratio alevin:\")\n",
    "print(\"Stewart et al.:\", fig1d.loc[(fig1d[\"Method\"] == \"Alevin-fry\") & (fig1d[\"Dataset\"] == \"Stewart et al.\"), \"Pct\"].mean())\n",
    "print(\"Mathew et al.:\", fig1d.loc[(fig1d[\"Method\"] == \"Alevin-fry\") & (fig1d[\"Dataset\"] == \"Mathew et al.\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (SE):\", fig1d.loc[(fig1d[\"Method\"] == \"Alevin-fry\") & (fig1d[\"Dataset\"] == \"Fu et al. (SE)\"), \"Pct\"].mean())\n",
    "print(\"Fu et al. (PE):\", fig1d.loc[(fig1d[\"Method\"] == \"Alevin-fry\") & (fig1d[\"Dataset\"] == \"Fu et al. (PE)\"), \"Pct\"].mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cf1d2e6-5604-43b6-bfa8-384c7bb4b1f0",
   "metadata": {},
   "source": [
    "### Figure 1 combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "161ea456-698d-45fe-a0bd-69dff8d32038",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up figure\n",
    "\n",
    "fig, axs = plt.subplots(1, 4, figsize=(12,2))\n",
    "\n",
    "# 1a\n",
    "plt.subplot(1,4,1)\n",
    "\n",
    "sns.boxplot(x=\"Dataset\", y=\"Counts\", hue=\"Method\", data=fig1a, \n",
    "            gap=0.3, legend=True, palette=[\"#548235\", \"#96C09F\", \"#6CA644\"], \n",
    "            log_scale=True, showfliers=False, width=0.8)\n",
    "plt.xlabel('')\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.ylabel(\"Counts per cell [log]\")\n",
    "\n",
    "plt.legend(bbox_to_anchor=(7.2,0.78))\n",
    "\n",
    "# 1b\n",
    "plt.subplot(1,4,2)\n",
    "\n",
    "sns.boxplot(x=\"Dataset\", y=\"Pct\", hue=\"Method\", data=fig1b, \n",
    "               gap=0.2, legend=False, palette=[\"#548235\", \"#96C09F\"], showfliers=False)\n",
    "plt.xlabel('')\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.ylabel(\"Count difference\")\n",
    "\n",
    "# 1c\n",
    "plt.subplot(1,4,3)\n",
    "\n",
    "sns.boxplot(x=\"Dataset\", y=\"Pearson correlation\", hue=\"Method\", data=fig1c, \n",
    "               gap=0.3, legend=False, palette=[\"#548235\", \"#96C09F\"], showfliers=False)\n",
    "plt.xlabel('')\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.ylabel(\"Pearson correlation\")\n",
    "   \n",
    "# 1d\n",
    "plt.subplot(1,4,4)\n",
    "\n",
    "sns.boxplot(x=\"Dataset\", y=\"Pct\", hue=\"Method\", data=fig1d, \n",
    "               gap=0.3, legend=False, palette=[\"#548235\", \"#96C09F\"], showfliers=False)\n",
    "plt.xlabel('')\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.ylabel(\"Unspliced counts [%]\")\n",
    "\n",
    "sns.despine()\n",
    "plt.subplots_adjust(wspace=0.7)\n",
    "\n",
    "plt.savefig(\"figures/1.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0038dd39-1928-4823-adf1-ad9f6a356373",
   "metadata": {},
   "source": [
    "## Figure 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7b00418-799b-4acc-ae5e-5a2856b88cc8",
   "metadata": {},
   "source": [
    "### Figure 2a: overlapping genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2f21762-0afc-4021-adcd-e436efcc3071",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stewart\n",
    "\n",
    "# get overlapping genes\n",
    "overlapping_genes_stewart = pd.read_pickle(\"../Stewart/data/stewart_overlapping_genes.pkl.gz\", compression='gzip')\n",
    "\n",
    "# subset: get only genes with mean expression >= 1\n",
    "subset_stewart = overlapping_genes_stewart[overlapping_genes_stewart[\"mean counts cellranger\"] >= 1]\n",
    "\n",
    "# sort by relative overlap\n",
    "subset_stewart.sort_values(by=\"rel_overlap\", ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0fead4e-e1fa-4096-aed0-f47a51f265cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(4, 4, figsize=(8,7))\n",
    "plt.suptitle(\"Correlation of overlapping genes\", size=14)\n",
    "\n",
    "for i in range(16):\n",
    "    plt.subplot(4,4,i+1)\n",
    "\n",
    "    gene_x = subset_stewart.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "\n",
    "    plt.scatter(x, y, s=0.5, c='grey')\n",
    "    \n",
    "    plt.xlabel(gene_x, size=8)\n",
    "    plt.ylabel(gene_y, size=8)\n",
    "    plt.xlim(left=0)\n",
    "    plt.ylim(bottom=0)\n",
    "    plt.tick_params(axis='x', labelsize=8)\n",
    "    plt.tick_params(axis='y', labelsize=8)\n",
    "    \n",
    "    # calculate Pearson correlation\n",
    "    corr, _ = pearsonr(x, y)\n",
    "\n",
    "    # display Pearson correlation and overlap\n",
    "    plt.text(0.25, 0.9, f\"r={corr:.2f}\", \n",
    "             size=8, ha='center', va='center', transform=plt.gca().transAxes);\n",
    "\n",
    "fig.supxlabel(\"Cell Ranger counts\", size=12)\n",
    "fig.supylabel(\"Velocyto unspliced counts\", size=12)\n",
    "\n",
    "plt.subplots_adjust(top=0.93, left=0.12, bottom=0.11, hspace=0.7, wspace=0.7)\n",
    "plt.savefig(\"figures/2a.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a1b016-c72f-4d27-807e-749d4ef88206",
   "metadata": {},
   "source": [
    "### Figure 2b: summary statistics of correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97ef352f-1c23-43a0-802f-d1547e962cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get overlapping genes of Mathew and Fu\n",
    "\n",
    "# Mathew\n",
    "\n",
    "# get overlapping genes\n",
    "overlapping_genes_mathew = pd.read_pickle(\"../Mathew/data/mathew_overlapping_genes.pkl.gz\", compression='gzip')\n",
    "\n",
    "# subset: get only genes with mean expression >= 1\n",
    "subset_mathew = overlapping_genes_mathew[overlapping_genes_mathew[\"mean counts cellranger\"] >= 1]\n",
    "\n",
    "# sort by relative overlap\n",
    "subset_mathew.sort_values(by=\"rel_overlap\", ascending=False, inplace=True)\n",
    "\n",
    "# Fu\n",
    "\n",
    "# get overlapping genes\n",
    "overlapping_genes_fu = pd.read_pickle(\"../Fu/data/fu_overlapping_genes.pkl.gz\", compression='gzip')\n",
    "\n",
    "# subset: get only genes with mean expression >= 1\n",
    "subset_fu = overlapping_genes_fu[overlapping_genes_fu[\"mean counts cellranger\"] >= 1]\n",
    "\n",
    "# sort by relative overlap\n",
    "subset_fu.sort_values(by=\"rel_overlap\", ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ea2b192-f5fc-45bd-ae53-e51a44c9cbe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Pearson correlations\n",
    "corr_stewart, corr_mathew, corr_fu, corr_fu_se, corr_fu_pe = [], [], [], [], []\n",
    "\n",
    "# Stewart\n",
    "for i in range(len(subset_stewart)):\n",
    "    gene_x = subset_stewart.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_stewart.append(corr)\n",
    "\n",
    "# Mathew\n",
    "for i in range(len(subset_mathew)):\n",
    "    gene_x = subset_mathew.index.get_level_values(0)[i]\n",
    "    gene_y = subset_mathew.index.get_level_values(1)[i]\n",
    "    x = mathew_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = mathew_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_mathew.append(corr)\n",
    "\n",
    "# Fu \n",
    "for i in range(len(subset_fu)):\n",
    "    gene_x = subset_fu.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu.index.get_level_values(1)[i]\n",
    "    x = fu_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu.append(corr)\n",
    "\n",
    "# Fu SE\n",
    "for i in range(len(subset_fu)):\n",
    "    gene_x = subset_fu.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu.index.get_level_values(1)[i]\n",
    "    x = fu_se_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_se_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_se.append(corr)\n",
    "\n",
    "# Fu PE\n",
    "for i in range(len(subset_fu)):\n",
    "    gene_x = subset_fu.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu.index.get_level_values(1)[i]\n",
    "    x = fu_pe_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_pe_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_pe.append(corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c60590b-45fc-4fb1-bf62-8056a4a5afd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get subsets with relative overlap > 60%\n",
    "subset_stewart_2 = subset_stewart[subset_stewart[\"rel_overlap\"] >= 0.6]\n",
    "subset_mathew_2 = subset_mathew[subset_mathew[\"rel_overlap\"] >= 0.6]\n",
    "subset_fu_2 = subset_fu[subset_fu[\"rel_overlap\"] >= 0.6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62d12b51-cb38-4f9e-b3cd-c1b286b06b48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Pearson correlations of subsets\n",
    "corr_stewart_2, corr_mathew_2, corr_fu_2, corr_fu_se_2, corr_fu_pe_2 = [], [], [], [], []\n",
    "\n",
    "# Stewart\n",
    "for i in range(len(subset_stewart_2)):\n",
    "    gene_x = subset_stewart_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart_2.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_stewart_2.append(corr)\n",
    "\n",
    "# Mathew\n",
    "for i in range(len(subset_mathew_2)):\n",
    "    gene_x = subset_mathew_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_mathew_2.index.get_level_values(1)[i]\n",
    "    x = mathew_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = mathew_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_mathew_2.append(corr)\n",
    "\n",
    "# Fu\n",
    "for i in range(len(subset_fu_2)):\n",
    "    gene_x = subset_fu_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_2.index.get_level_values(1)[i]\n",
    "    x = fu_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_2.append(corr)\n",
    "\n",
    "# Fu (SE)\n",
    "for i in range(len(subset_fu_2)):\n",
    "    gene_x = subset_fu_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_2.index.get_level_values(1)[i]\n",
    "    x = fu_se_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_se_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_se_2.append(corr)\n",
    "\n",
    "# Fu (PE)\n",
    "for i in range(len(subset_fu_2)):\n",
    "    gene_x = subset_fu_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_2.index.get_level_values(1)[i]\n",
    "    x = fu_pe_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_pe_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_pe_2.append(corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e87c2b89-8ad2-487e-ba83-5bb75b3f6e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get subsets with relative overlap > 80%\n",
    "subset_stewart_3 = subset_stewart[subset_stewart[\"rel_overlap\"] >= 0.8]\n",
    "subset_mathew_3 = subset_mathew[subset_mathew[\"rel_overlap\"] >= 0.8]\n",
    "subset_fu_3 = subset_fu[subset_fu[\"rel_overlap\"] >= 0.8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a86c9f3-7fc6-47c3-85c5-418cb940f02e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Pearson correlations of subsets\n",
    "corr_stewart_3, corr_mathew_3, corr_fu_3, corr_fu_se_3, corr_fu_pe_3 = [], [], [], [], []\n",
    "\n",
    "# Stewart\n",
    "for i in range(len(subset_stewart_3)):\n",
    "    gene_x = subset_stewart_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart_3.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_stewart_3.append(corr)\n",
    "\n",
    "# Mathew\n",
    "for i in range(len(subset_mathew_3)):\n",
    "    gene_x = subset_mathew_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_mathew_3.index.get_level_values(1)[i]\n",
    "    x = mathew_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = mathew_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_mathew_3.append(corr)\n",
    "\n",
    "# Fu\n",
    "for i in range(len(subset_fu_3)):\n",
    "    gene_x = subset_fu_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_3.index.get_level_values(1)[i]\n",
    "    x = fu_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_3.append(corr)\n",
    "\n",
    "# Fu (SE)\n",
    "for i in range(len(subset_fu_3)):\n",
    "    gene_x = subset_fu_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_3.index.get_level_values(1)[i]\n",
    "    x = fu_se_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_se_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_se_3.append(corr)\n",
    "\n",
    "# Fu (PE)\n",
    "for i in range(len(subset_fu_3)):\n",
    "    gene_x = subset_fu_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_3.index.get_level_values(1)[i]\n",
    "    x = fu_pe_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_pe_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_pe_3.append(corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8712d952-e6f8-4241-8d3f-4232fa3ffbac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe to plot\n",
    "\n",
    "fig2b = pd.DataFrame({\"Pearson correlation\": np.concatenate([corr_stewart, corr_stewart_2, corr_stewart_3,\n",
    "                                                             corr_mathew, corr_mathew_2, corr_mathew_3,\n",
    "                                                             corr_fu_se, corr_fu_se_2, corr_fu_se_3,\n",
    "                                                             corr_fu_pe, corr_fu_pe_2, corr_fu_pe_3]),\n",
    "                      \"Subset\": np.concatenate([[\"All genes\"] * len(subset_stewart), [\">60% overlap\"] * len(subset_stewart_2), [\">80% overlap\"] * len(subset_stewart_3),\n",
    "                                                [\"All genes\"] * len(subset_mathew), [\">60% overlap\"] * len(subset_mathew_2), [\">80% overlap\"] * len(subset_mathew_3),\n",
    "                                                [\"All genes\"] * len(subset_fu), [\">60% overlap\"] * len(subset_fu_2), [\">80% overlap\"] * len(subset_fu_3),\n",
    "                                                [\"All genes\"] * len(subset_fu), [\">60% overlap\"] * len(subset_fu_2), [\">80% overlap\"] * len(subset_fu_3)]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * (len(subset_stewart) + len(subset_stewart_2) + len(subset_stewart_3)),\n",
    "                                                 [\"Mathew et al.\"] * (len(subset_mathew) + len(subset_mathew_2) + len(subset_mathew_3)),\n",
    "                                                 [\"Fu et al. (SE)\"] * (len(subset_fu) + len(subset_fu_2) + len(subset_fu_3)),\n",
    "                                                 [\"Fu et al. (PE)\"] * (len(subset_fu) + len(subset_fu_2) + len(subset_fu_3))])})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "175df6a6-6dde-445e-9e39-4b38889bda38",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 2B: summary of correlations of overlapping gene counts\n",
    "\n",
    "# set up figure\n",
    "fig, axs = plt.subplots(1, 1, figsize=(6,4))\n",
    "plt.title(\"Correlation of overlapping genes\", size=14)\n",
    "sns.boxplot(x=\"Dataset\", y=\"Pearson correlation\", hue=\"Subset\", data=fig2b, \n",
    "               gap=0.6, showfliers=False, palette=[\"#548235\", \"#96C09F\", \"#6CA644\"])\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.legend(bbox_to_anchor=(1.5,0.6))\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "sns.despine()\n",
    "plt.savefig(\"figures/2b.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef80bea6-8a46-42e9-9482-35af8e7ee72d",
   "metadata": {},
   "source": [
    "## Supplement 1: overlapping genes Mathew"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b19f0783-2b38-4b14-9201-e0bc67450c90",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(4, 4, figsize=(8,7))\n",
    "plt.suptitle(\"Correlation of overlapping genes\", size=14)\n",
    "\n",
    "for i in range(16):\n",
    "    plt.subplot(4,4,i+1)\n",
    "\n",
    "    gene_x = subset_mathew.index.get_level_values(0)[i]\n",
    "    gene_y = subset_mathew.index.get_level_values(1)[i]\n",
    "    x = mathew_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = mathew_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "\n",
    "    plt.scatter(x, y, s=0.5, c='grey')\n",
    "    \n",
    "    plt.xlabel(gene_x, size=8)\n",
    "    plt.ylabel(gene_y, size=8)\n",
    "    plt.xlim(left=0)\n",
    "    plt.ylim(bottom=0)\n",
    "    plt.tick_params(axis='x', labelsize=8)\n",
    "    plt.tick_params(axis='y', labelsize=8)\n",
    "    \n",
    "    # calculate Pearson correlation\n",
    "    corr, _ = pearsonr(x, y)\n",
    "\n",
    "    # display Pearson correlation and overlap\n",
    "    plt.text(0.25, 0.9, f\"r={corr:.2f}\", \n",
    "             size=8, ha='center', va='center', transform=plt.gca().transAxes);\n",
    "\n",
    "fig.supxlabel(\"Cellranger counts\", size=12)\n",
    "fig.supylabel(\"Velocyto unspliced counts\", size=12)\n",
    "\n",
    "plt.subplots_adjust(top=0.93, left=0.12, bottom=0.11, hspace=0.7, wspace=0.7)\n",
    "plt.savefig(\"figures/supplement_1.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60cfe299-b709-4cfd-a2b2-aae82bbf7b84",
   "metadata": {},
   "source": [
    "## Supplement 2: overlapping genes Fu SE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8fb3464-bd72-4929-8771-9de5875b7bf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(4, 4, figsize=(8,7))\n",
    "plt.suptitle(\"Correlation of overlapping genes\", size=14)\n",
    "\n",
    "for i in range(16):\n",
    "    plt.subplot(4,4,i+1)\n",
    "\n",
    "    gene_x = subset_fu.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu.index.get_level_values(1)[i]\n",
    "    x = fu_se_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_se_velocyto[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "\n",
    "    plt.scatter(x, y, s=0.5, c='grey')\n",
    "    \n",
    "    plt.xlabel(gene_x, size=8)\n",
    "    plt.ylabel(gene_y, size=8)\n",
    "    plt.xlim(left=0)\n",
    "    plt.ylim(bottom=0)\n",
    "    plt.tick_params(axis='x', labelsize=8)\n",
    "    plt.tick_params(axis='y', labelsize=8)\n",
    "    \n",
    "    # calculate Pearson correlation\n",
    "    corr, _ = pearsonr(x, y)\n",
    "\n",
    "    # display Pearson correlation and overlap\n",
    "    plt.text(0.25, 0.9, f\"r={corr:.2f}\", \n",
    "             size=8, ha='center', va='center', transform=plt.gca().transAxes);\n",
    "\n",
    "fig.supxlabel(\"Cellranger counts\", size=12)\n",
    "fig.supylabel(\"Velocyto unspliced counts\", size=12)\n",
    "\n",
    "plt.subplots_adjust(top=0.93, left=0.12, bottom=0.11, hspace=0.7, wspace=0.7)\n",
    "plt.savefig(\"figures/supplement_2.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfe25622-4f53-4bc5-883e-03025f9ea686",
   "metadata": {},
   "source": [
    "## Supplement 3: co-expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "411dd6e1-77a7-4848-912c-f2e758435c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(4, 4, figsize=(8,7))\n",
    "plt.suptitle(\"Correlation of overlapping genes\", size=14)\n",
    "\n",
    "for i in range(16):\n",
    "    plt.subplot(4,4,i+1)\n",
    "\n",
    "    gene_x = subset_stewart.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_velocyto[:,gene_y].layers[\"counts\"].toarray().ravel()\n",
    "\n",
    "    plt.scatter(x, y, s=0.5, c='grey')\n",
    "    \n",
    "    plt.xlabel(gene_x, size=8)\n",
    "    plt.ylabel(gene_y, size=8)\n",
    "    plt.xlim(left=0)\n",
    "    plt.ylim(bottom=0)\n",
    "    plt.tick_params(axis='x', labelsize=8)\n",
    "    plt.tick_params(axis='y', labelsize=8)\n",
    "    \n",
    "    # calculate Pearson correlation\n",
    "    corr, _ = pearsonr(x, y)\n",
    "\n",
    "    # display Pearson correlation and overlap\n",
    "    plt.text(0.25, 0.9, f\"r={corr:.2f}\", \n",
    "             size=8, ha='center', va='center', transform=plt.gca().transAxes);\n",
    "\n",
    "fig.supxlabel(\"Cellranger counts\", size=12)\n",
    "fig.supylabel(\"Cellranger counts\", size=12)\n",
    "\n",
    "plt.subplots_adjust(top=0.93, left=0.12, bottom=0.11, hspace=0.7, wspace=0.7)\n",
    "plt.savefig(\"figures/supplement_3.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e67d659-6510-4b7e-a6bb-1acbfa141175",
   "metadata": {},
   "source": [
    "## Supplement 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c3aa738-d530-4a0b-aae0-cb4c1598df94",
   "metadata": {},
   "source": [
    "### Supplement 4a: correlation in alevin-fry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b956277-f5fd-48ad-8cc5-9f634917e62a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(4, 4, figsize=(8,7))\n",
    "plt.suptitle(\"Correlation of overlapping genes\", size=14)\n",
    "\n",
    "for i in range(16):\n",
    "    plt.subplot(4,4,i+1)\n",
    "\n",
    "    gene_x = subset_stewart.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "\n",
    "    plt.scatter(x, y, s=0.5, c='grey')\n",
    "    \n",
    "    plt.xlabel(gene_x, size=8)\n",
    "    plt.ylabel(gene_y, size=8)\n",
    "    plt.xlim(left=0)\n",
    "    plt.ylim(bottom=0)\n",
    "    plt.tick_params(axis='x', labelsize=8)\n",
    "    plt.tick_params(axis='y', labelsize=8)\n",
    "    \n",
    "    # calculate Pearson correlation\n",
    "    corr, _ = pearsonr(x, y)\n",
    "\n",
    "    # display Pearson correlation and overlap\n",
    "    plt.text(0.25, 0.9, f\"r={corr:.2f}\", \n",
    "             size=8, ha='center', va='center', transform=plt.gca().transAxes);\n",
    "\n",
    "fig.supxlabel(\"Cellranger counts\", size=12)\n",
    "fig.supylabel(\"Alevin-fry unspliced counts\", size=12)\n",
    "\n",
    "plt.subplots_adjust(top=0.93, left=0.12, bottom=0.11, hspace=0.7, wspace=0.7)\n",
    "plt.savefig(\"figures/supplement_4a.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60dd38f6-5a4f-43bc-889c-181991fd9358",
   "metadata": {},
   "source": [
    "### Supplement 4b: summary statistics of correlations in alevin-fry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c98dea9-0604-475c-8bce-b595b6e96f3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Pearson correlations\n",
    "corr_stewart_alevin, corr_mathew_alevin, corr_fu_se_alevin, corr_fu_pe_alevin = [], [], [], []\n",
    "\n",
    "# Stewart\n",
    "for i in range(len(subset_stewart)):\n",
    "    gene_x = subset_stewart.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_stewart_alevin.append(corr)\n",
    "\n",
    "# Mathew\n",
    "for i in range(len(subset_mathew)):\n",
    "    gene_x = subset_mathew.index.get_level_values(0)[i]\n",
    "    gene_y = subset_mathew.index.get_level_values(1)[i]\n",
    "    x = mathew_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = mathew_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_mathew_alevin.append(corr)\n",
    "\n",
    "# Fu SE\n",
    "for i in range(len(subset_fu)):\n",
    "    gene_x = subset_fu.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu.index.get_level_values(1)[i]\n",
    "    x = fu_se_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_se_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_se_alevin.append(corr)\n",
    "\n",
    "# Fu PE\n",
    "for i in range(len(subset_fu)):\n",
    "    gene_x = subset_fu.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu.index.get_level_values(1)[i]\n",
    "    x = fu_pe_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_pe_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_pe_alevin.append(corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cd097ff-5f2c-40a9-af2a-865bda3cf0c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Pearson correlations of subsets\n",
    "corr_stewart_alevin_2, corr_mathew_alevin_2, corr_fu_se_alevin_2, corr_fu_pe_alevin_2 = [], [], [], []\n",
    "\n",
    "# Stewart\n",
    "for i in range(len(subset_stewart_2)):\n",
    "    gene_x = subset_stewart_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart_2.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_stewart_alevin_2.append(corr)\n",
    "\n",
    "# Mathew\n",
    "for i in range(len(subset_mathew_2)):\n",
    "    gene_x = subset_mathew_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_mathew_2.index.get_level_values(1)[i]\n",
    "    x = mathew_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = mathew_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_mathew_alevin_2.append(corr)\n",
    "\n",
    "# Fu (SE)\n",
    "for i in range(len(subset_fu_2)):\n",
    "    gene_x = subset_fu_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_2.index.get_level_values(1)[i]\n",
    "    x = fu_se_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_se_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_se_alevin_2.append(corr)\n",
    "\n",
    "# Fu (PE)\n",
    "for i in range(len(subset_fu_2)):\n",
    "    gene_x = subset_fu_2.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_2.index.get_level_values(1)[i]\n",
    "    x = fu_pe_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_pe_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_pe_alevin_2.append(corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2255ef13-7c90-4da4-adbb-53c2d812ffa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Pearson correlations of subsets\n",
    "corr_stewart_alevin_3, corr_mathew_alevin_3, corr_fu_se_alevin_3, corr_fu_pe_alevin_3 = [], [], [], []\n",
    "\n",
    "# Stewart\n",
    "for i in range(len(subset_stewart_3)):\n",
    "    gene_x = subset_stewart_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_stewart_3.index.get_level_values(1)[i]\n",
    "    x = stewart_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = stewart_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_stewart_alevin_3.append(corr)\n",
    "\n",
    "# Mathew\n",
    "for i in range(len(subset_mathew_3)):\n",
    "    gene_x = subset_mathew_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_mathew_3.index.get_level_values(1)[i]\n",
    "    x = mathew_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = mathew_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_mathew_alevin_3.append(corr)\n",
    "\n",
    "# Fu (SE)\n",
    "for i in range(len(subset_fu_3)):\n",
    "    gene_x = subset_fu_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_3.index.get_level_values(1)[i]\n",
    "    x = fu_se_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_se_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_se_alevin_3.append(corr)\n",
    "\n",
    "# Fu (PE)\n",
    "for i in range(len(subset_fu_3)):\n",
    "    gene_x = subset_fu_3.index.get_level_values(0)[i]\n",
    "    gene_y = subset_fu_3.index.get_level_values(1)[i]\n",
    "    x = fu_pe_velocyto[:,gene_x].layers[\"counts\"].toarray().ravel()\n",
    "    y = fu_pe_alevin[:,gene_y].layers[\"unspliced\"].toarray().ravel()\n",
    "    corr, _ = pearsonr(x, y)\n",
    "    corr_fu_pe_alevin_3.append(corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d28c309-f6bb-46dc-9daa-3c0bf51643b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe to plot\n",
    "\n",
    "sup4b = pd.DataFrame({\"Pearson correlation\": np.concatenate([corr_stewart_alevin, corr_stewart_alevin_2, corr_stewart_alevin_3,\n",
    "                                                             corr_mathew_alevin, corr_mathew_alevin_2, corr_mathew_alevin_3,\n",
    "                                                             corr_fu_se_alevin, corr_fu_se_alevin_2, corr_fu_se_alevin_3,\n",
    "                                                             corr_fu_pe_alevin, corr_fu_pe_alevin_2, corr_fu_pe_alevin_3]),\n",
    "                      \"Subset\": np.concatenate([[\"All genes\"] * len(subset_stewart), [\">60% overlap\"] * len(subset_stewart_2), [\">80% overlap\"] * len(subset_stewart_3),\n",
    "                                                [\"All genes\"] * len(subset_mathew), [\">60% overlap\"] * len(subset_mathew_2), [\">80% overlap\"] * len(subset_mathew_3),\n",
    "                                                [\"All genes\"] * len(subset_fu), [\">60% overlap\"] * len(subset_fu_2), [\">80% overlap\"] * len(subset_fu_3),\n",
    "                                                [\"All genes\"] * len(subset_fu), [\">60% overlap\"] * len(subset_fu_2), [\">80% overlap\"] * len(subset_fu_3)]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * (len(subset_stewart) + len(subset_stewart_2) + len(subset_stewart_3)),\n",
    "                                                 [\"Mathew et al.\"] * (len(subset_mathew) + len(subset_mathew_2) + len(subset_mathew_3)),\n",
    "                                                 [\"Fu et al. (SE)\"] * (len(subset_fu) + len(subset_fu_2) + len(subset_fu_3)),\n",
    "                                                 [\"Fu et al. (PE)\"] * (len(subset_fu) + len(subset_fu_2) + len(subset_fu_3))])})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af8ba5fc-b366-45fe-b0df-3eb870327a46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure S4B: summary of correlations of overlapping gene counts\n",
    "\n",
    "# set up figure\n",
    "fig, axs = plt.subplots(1, 1, figsize=(6,4))\n",
    "plt.title(\"Correlation of overlapping genes\", size=14)\n",
    "sns.boxplot(x=\"Dataset\", y=\"Pearson correlation\", hue=\"Subset\", data=sup4b, \n",
    "               gap=0.6, showfliers=False, palette=[\"#548235\", \"#96C09F\", \"#6CA644\"])\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.legend(bbox_to_anchor=(1.5,0.6))\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "sns.despine()\n",
    "plt.savefig(\"figures/supplement_4b.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd1c7a55-c434-4436-a8b4-d3cfb10a3478",
   "metadata": {},
   "source": [
    "## Supplement 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e4cae0c-d5f8-48ff-b833-4e010abf2625",
   "metadata": {},
   "source": [
    "### Supplement 5a: coorelation of spliced counts: MIF (overlaps with exon only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "408cf631-80ea-40ee-9705-fd243e3ac006",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(2, 1, figsize=(2,3.7))\n",
    "\n",
    "# Stewart\n",
    "plt.subplot(2,1,1)\n",
    "    \n",
    "x = stewart_velocyto[:,\"MIF\"].layers[\"counts\"].toarray().ravel()\n",
    "y = stewart_velocyto[:,\"MIF-AS1\"].layers[\"spliced\"].toarray().ravel()\n",
    "\n",
    "plt.scatter(x, y, s=0.5, c='grey')\n",
    "    \n",
    "plt.xlabel(\"MIF\", size=8)\n",
    "plt.ylabel(\"MIF-AS1\", size=8)\n",
    "plt.xlim(left=0)\n",
    "plt.ylim(bottom=0)\n",
    "plt.tick_params(axis='x', labelsize=8)\n",
    "plt.tick_params(axis='y', labelsize=8)\n",
    "    \n",
    "# calculate Pearson correlation\n",
    "corr, _ = pearsonr(x, y)\n",
    "\n",
    "# display Pearson correlation\n",
    "plt.text(0.25, 0.9, f\"r={corr:.2f}\", size=8, ha='center', va='center', transform=plt.gca().transAxes)\n",
    "\n",
    "# Fu\n",
    "plt.subplot(2,1,2)\n",
    "    \n",
    "x = fu_se_velocyto[:,\"MIF\"].layers[\"counts\"].toarray().ravel()\n",
    "y = fu_se_velocyto[:,\"MIF-AS1\"].layers[\"spliced\"].toarray().ravel()\n",
    "\n",
    "plt.scatter(x, y, s=0.5, c='grey')\n",
    "    \n",
    "plt.xlabel(\"MIF\", size=8)\n",
    "plt.ylabel(\"MIF-AS1\", size=8)\n",
    "plt.xlim(left=0)\n",
    "plt.ylim(bottom=0)\n",
    "plt.tick_params(axis='x', labelsize=8)\n",
    "plt.tick_params(axis='y', labelsize=8)\n",
    "    \n",
    "# calculate Pearson correlation\n",
    "corr, _ = pearsonr(x, y)\n",
    "\n",
    "# display Pearson correlation\n",
    "plt.text(0.25, 0.9, f\"r={corr:.2f}\", size=8, ha='center', va='center', transform=plt.gca().transAxes)\n",
    "\n",
    "fig.supxlabel(\"Cellranger counts\", size=12)\n",
    "fig.supylabel(\"Velocyto spliced counts\", size=12)\n",
    "\n",
    "plt.subplots_adjust(left=0.4, hspace=0.6, bottom=0.2)\n",
    "plt.savefig(\"figures/supplement_5a.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d661ea-183b-4607-b678-2e7d46eaa9b7",
   "metadata": {},
   "source": [
    "## Figure 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3acad31a-9258-4b17-aaca-2f9484ef6997",
   "metadata": {},
   "source": [
    "### Figure 3a: Velocity gene overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03159a5b-d8ac-411b-85b6-f7087eefd9fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get overlap\n",
    "velocity_genes = pd.DataFrame({\"velocyto\": [np.sum(stewart_velocyto.var[\"velocity_genes\"]),\n",
    "                                            np.sum(mathew_velocyto.var[\"velocity_genes\"]),\n",
    "                                            np.sum(fu_velocyto.var[\"velocity_genes\"])],\n",
    "                               \"alevin\": [np.sum(stewart_alevin.var[\"velocity_genes\"]),\n",
    "                                          np.sum(mathew_alevin.var[\"velocity_genes\"]),\n",
    "                                          np.sum(fu_alevin.var[\"velocity_genes\"])],\n",
    "                               \"overlap\": [len(set(stewart_velocyto[:,stewart_velocyto.var[\"velocity_genes\"] == True].var_names) & \n",
    "                                               set(stewart_alevin[:,stewart_alevin.var[\"velocity_genes\"] == True].var_names)),\n",
    "                                           len(set(mathew_velocyto[:,mathew_velocyto.var[\"velocity_genes\"] == True].var_names) & \n",
    "                                               set(mathew_alevin[:,mathew_alevin.var[\"velocity_genes\"] == True].var_names)),\n",
    "                                           len(set(fu_velocyto[:,fu_velocyto.var[\"velocity_genes\"] == True].var_names) & \n",
    "                                               set(fu_alevin[:,fu_alevin.var[\"velocity_genes\"] == True].var_names))]},\n",
    "                              index = [\"Stewart et al.\", \"Mathew et al.\", \"Fu et al.\"])                                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3d96d4c-98d5-4945-a42a-a4c562c42e52",
   "metadata": {},
   "outputs": [],
   "source": [
    "velocity_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9202d7e7-cee0-4fd7-aeed-a81fe108b642",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up figure\n",
    "fig, axs = plt.subplots(1, 3, figsize=(8,4))\n",
    "\n",
    "plt.subplot(1,3,1)\n",
    "out = venn2(subsets = (24, 200, 23), set_colors=[\"#548235\", \"#96C09F\"], set_labels=[\"\",\"\"], alpha=1)\n",
    "for text in out.subset_labels:\n",
    "   text.set_fontsize(10)\n",
    "plt.ylabel(\"Stewart et al.\")\n",
    "\n",
    "plt.subplot(1,3,2)\n",
    "out = venn2(subsets = (24, 683, 103), set_colors=[\"#548235\", \"#96C09F\"], set_labels=[\"\",\"\"], alpha=1)\n",
    "for text in out.subset_labels:\n",
    "   text.set_fontsize(10)\n",
    "plt.ylabel(\"Mathew et al.\")\n",
    "\n",
    "plt.subplot(1,3,3)\n",
    "out = venn2(subsets = (73, 296, 152), set_colors=[\"#548235\", \"#96C09F\"], set_labels=[\"\",\"\"], alpha=1)\n",
    "for text in out.subset_labels:\n",
    "   text.set_fontsize(10)\n",
    "plt.ylabel(\"Fu et al.\")\n",
    "\n",
    "plt.savefig(\"figures/3a.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c02d78a-cdaa-4a94-8bb9-9516529c4d08",
   "metadata": {},
   "source": [
    "### Figure 3b: phase plots of marker genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b820060-febb-415c-9a47-5f0679d1e60d",
   "metadata": {},
   "source": [
    "#### determine 5% and 95% percentiles to color datapoints used for steady-state ratio calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13fa7249-54c7-4d25-b549-b8570128600c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stewart\n",
    "\n",
    "# velocyto\n",
    "lr = LinearRegression(fit_intercept=False, percentile=[5, 95])\n",
    "n_obs, x, y = lr._trim_data(data=[stewart_velocyto[:,\"TCL1A\"].layers[\"Ms\"], stewart_velocyto[:,\"TCL1A\"].layers[\"Mu\"]])\n",
    "\n",
    "velo_quantiles = [i for i in range(stewart_velocyto.n_obs) if i in np.nonzero(x.toarray())[0] or i in np.nonzero(y.toarray())[0]]\n",
    "\n",
    "lr.fit(x,y)\n",
    "print(lr.coef_)\n",
    "\n",
    "# alevin\n",
    "lr = LinearRegression(fit_intercept=False, percentile=[5, 95])\n",
    "n_obs, x, y = lr._trim_data(data=[stewart_alevin[:,\"TCL1A\"].layers[\"Ms\"], stewart_alevin[:,\"TCL1A\"].layers[\"Mu\"]])\n",
    "\n",
    "alevin_quantiles = [i for i in range(stewart_alevin.n_obs) if i in np.nonzero(x.toarray())[0] or i in np.nonzero(y.toarray())[0]]\n",
    "\n",
    "lr.fit(x,y)\n",
    "print(lr.coef_)\n",
    "\n",
    "stewart_velocyto.obs[\"extreme_quantile\"] = [\"Both\" if i in velo_quantiles and i in alevin_quantiles\n",
    "                                                else \"Velocyto\" if i in velo_quantiles\n",
    "                                                else \"Alevin-fry\" if i in alevin_quantiles\n",
    "                                                else \"None\" for i in range(stewart_velocyto_sto.n_obs)]\n",
    "stewart_alevin.obs[\"extreme_quantile\"] = stewart_velocyto_sto.obs[\"extreme_quantile\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c694657-5b16-4b68-b6d0-4935426d51d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mathew\n",
    "\n",
    "# velocyto\n",
    "lr = LinearRegression(fit_intercept=False, percentile=[5, 95])\n",
    "n_obs, x, y = lr._trim_data(data=[mathew_velocyto[:,\"Mki67\"].layers[\"Ms\"], mathew_velocyto[:,\"Mki67\"].layers[\"Mu\"]])\n",
    "\n",
    "velo_quantiles = [i for i in range(mathew_velocyto.n_obs) if (i in np.nonzero(x.toarray())[0] or i in np.nonzero(y.toarray())[0])]\n",
    "\n",
    "lr.fit(x,y)\n",
    "print(lr.coef_)\n",
    "\n",
    "# alevin\n",
    "lr = LinearRegression(fit_intercept=False, percentile=[5, 95])\n",
    "n_obs, x, y = lr._trim_data(data=[mathew_alevin[:,\"Mki67\"].layers[\"Ms\"], mathew_alevin[:,\"Mki67\"].layers[\"Mu\"]])\n",
    "alevin_quantiles = [i for i in range(mathew_alevin.n_obs) if (i in np.nonzero(x.toarray())[0] or i in np.nonzero(y.toarray())[0])]\n",
    "\n",
    "lr.fit(x,y)\n",
    "print(lr.coef_)\n",
    "\n",
    "mathew_velocyto.obs[\"extreme_quantile\"] = [\"Both\" if i in velo_quantiles and i in alevin_quantiles\n",
    "                                                else \"Velocyto\" if i in velo_quantiles\n",
    "                                                else \"Alevin-fry\" if i in alevin_quantiles\n",
    "                                                else \"None\" for i in range(mathew_velocyto_sto.n_obs)]\n",
    "mathew_alevin.obs[\"extreme_quantile\"] = mathew_velocyto_sto.obs[\"extreme_quantile\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1424a07e-074d-4797-ab82-c4cc08b7bd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fu\n",
    "\n",
    "# velocyto\n",
    "lr = LinearRegression(fit_intercept=False, percentile=[5, 95])\n",
    "n_obs, x, y = lr._trim_data(data=[fu_velocyto[:,\"TBX21\"].layers[\"Ms\"], fu_velocyto[:,\"TBX21\"].layers[\"Mu\"]])\n",
    "\n",
    "velo_quantiles = [i for i in range(fu_velocyto.n_obs) if i in np.nonzero(x.toarray())[0] or i in np.nonzero(y.toarray())[0]]\n",
    "\n",
    "lr.fit(x,y)\n",
    "print(lr.coef_)\n",
    "\n",
    "# alevin\n",
    "lr = LinearRegression(fit_intercept=False, percentile=[5, 95])\n",
    "n_obs, x, y = lr._trim_data(data=[fu_alevin[:,\"TBX21\"].layers[\"Ms\"], fu_alevin[:,\"TBX21\"].layers[\"Mu\"]])\n",
    "alevin_quantiles = [i for i in range(fu_alevin.n_obs) if i in np.nonzero(x.toarray())[0] or i in np.nonzero(y.toarray())[0]]\n",
    "\n",
    "lr.fit(x,y)\n",
    "print(lr.coef_)\n",
    "\n",
    "fu_velocyto.obs[\"extreme_quantile\"] = [\"Both\" if i in velo_quantiles and i in alevin_quantiles\n",
    "                                            else \"Velocyto\" if i in velo_quantiles\n",
    "                                            else \"Alevin-fry\" if i in alevin_quantiles\n",
    "                                            else \"None\" for i in range(fu_velocyto_sto.n_obs)]\n",
    "fu_alevin.obs[\"extreme_quantile\"] = fu_velocyto_sto.obs[\"extreme_quantile\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aae9ac8-7711-45b4-a848-f0051dbae181",
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure 3b: phase plots of marker genes\n",
    "\n",
    "# set up figure\n",
    "fig, axs = plt.subplots(3, 2, figsize=(4,6))\n",
    "\n",
    "# Stewart\n",
    "\n",
    "scv.pl.scatter(stewart_velocyto, 'TCL1A', color=\"extreme_quantile\",\n",
    "               palette= {\"Both\": \"#d62728\", \"Velocyto\": \"#1f77b4\", \"Alevin-fry\": \"#2ca02c\", \"None\": \"lightgrey\"}, cmap=\"coolwarm\", \n",
    "               ax=axs[0][0], show=False, figsize=(3,3), xlabel=\"\", ylabel=\"TCL1A\\n\", title=\"Velocyto\\n\", size=5, fontsize=12)\n",
    "axs[0][0].get_legend().remove()\n",
    "\n",
    "scv.pl.scatter(stewart_alevin, 'TCL1A', color=\"extreme_quantile\",\n",
    "               palette= {\"Both\": \"#d62728\", \"Velocyto\": \"#1f77b4\", \"Alevin-fry\": \"#2ca02c\", \"None\": \"lightgrey\"}, cmap=\"coolwarm\",\n",
    "               ax=axs[0][1], show=False, figsize=(3,3), xlabel=\"\", ylabel=\"\", title=\"Alevin-fry\\n\", size=5, fontsize=12)\n",
    "axs[0][1].get_legend().remove()\n",
    "\n",
    "# Mathew\n",
    "\n",
    "scv.pl.scatter(mathew_velocyto, 'Mki67', color=\"extreme_quantile\",\n",
    "               palette= {\"Both\": \"#d62728\", \"Velocyto\": \"#1f77b4\", \"Alevin-fry\": \"#2ca02c\", \"None\": \"lightgrey\"}, cmap=\"coolwarm\", \n",
    "               ax=axs[1][0], show=False, figsize=(3,3), xlabel=\"\", ylabel=\"Mki67\\n\", title=\"\", size=5, fontsize=12)\n",
    "axs[1][0].get_legend().remove()\n",
    "\n",
    "scv.pl.scatter(mathew_alevin, 'Mki67', color=\"extreme_quantile\",\n",
    "               palette= {\"Both\": \"#d62728\", \"Velocyto\": \"#1f77b4\", \"Alevin-fry\": \"#2ca02c\", \"None\": \"lightgrey\"}, cmap=\"coolwarm\",\n",
    "               ax=axs[1][1], show=False, figsize=(3,3), xlabel=\"\", ylabel=\"\", title=\"\", size=5)\n",
    "axs[1][1].get_legend().remove()\n",
    "\n",
    "#Fu\n",
    "\n",
    "scv.pl.scatter(fu_velocyto, 'TBX21', color=\"extreme_quantile\",\n",
    "               palette= {\"Both\": \"#d62728\", \"Velocyto\": \"#1f77b4\", \"Alevin-fry\": \"#2ca02c\", \"None\": \"lightgrey\"}, cmap=\"coolwarm\",\n",
    "               ax=axs[2][0], show=False, figsize=(3,3), xlabel=\"\", ylabel=\"TBX21\\n\", title=\"\", size=5, fontsize=12)\n",
    "axs[2][0].get_legend().remove()\n",
    "\n",
    "scv.pl.scatter(fu_alevin, 'TBX21', color=\"extreme_quantile\",\n",
    "               palette= {\"Both\": \"#d62728\", \"Velocyto\": \"#1f77b4\", \"Alevin-fry\": \"#2ca02c\", \"None\": \"lightgrey\"}, cmap=\"coolwarm\",\n",
    "               ax=axs[2][1], show=False, figsize=(3,3), xlabel=\"\", ylabel=\"\", title=\"\", size=5)\n",
    "axs[2][1].get_legend().remove()\n",
    "\n",
    "# adjust spacing and save\n",
    "plt.subplots_adjust(wspace=0.5, hspace=0.5, left=0.12, top=0.85)\n",
    "plt.savefig(\"figures/3_b.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b417f1c2-37e6-4ac9-a2e6-7473187a2232",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add legend\n",
    "fig, ax = plt.subplots(figsize=(2,2))\n",
    "\n",
    "line1, = ax.plot([], [], marker=\"o\", linestyle='None', color=\"#d62728\", label='Both')\n",
    "line2, = ax.plot([], [], marker=\"o\", linestyle='None',  color=\"#1f77b4\", label='Velocyto')\n",
    "line3, = ax.plot([], [], marker=\"o\", linestyle='None', color=\"#2ca02c\", label='Alevin-fry')\n",
    "line4, = ax.plot([], [], marker=\"o\", linestyle='None', color=\"lightgrey\", label='None')\n",
    "\n",
    "ax.legend(title=\"Extreme quantile \\n\", title_fontsize=12, fontsize=12)\n",
    "ax.set_axis_off()\n",
    "\n",
    "plt.savefig(\"figures/3_b_legend.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "849824c4-8064-412e-8b8c-dc6d2bc096b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure 3c: velocities of marker genes\n",
    "\n",
    "# set up figure\n",
    "fig, axs = plt.subplots(3, 2, figsize=(6,8))\n",
    "\n",
    "# Stewart\n",
    "\n",
    "scv.pl.scatter(stewart_velocyto, layer=\"velocity\", color='TCL1A', size=4, vmin=-0.05, vmax=0.05, \n",
    "                cmap=\"coolwarm\", ax=axs[0][0], show=False, figsize=(3,3), title=\"Velocyto\\n\", colorbar=False, fontsize=12)\n",
    "\n",
    "scv.pl.scatter(stewart_alevin, layer=\"velocity\", color='TCL1A', size=4, vmin=-0.05, vmax=0.05,\n",
    "                cmap=\"coolwarm\", ax=axs[0][1], show=False, figsize=(3,3), title=\"Alevin-fry\\n\", colorbar=False, fontsize=12)\n",
    "\n",
    "# Mathew\n",
    "\n",
    "scv.pl.scatter(mathew_velocyto, layer=\"velocity\", color='Mki67', size=4, vmin=-0.05, vmax=0.05, \n",
    "                cmap=\"coolwarm\", ax=axs[1][0], show=False, figsize=(3,3), title=\"\", colorbar=False)\n",
    "\n",
    "scv.pl.scatter(mathew_alevin, layer=\"velocity\", color='Mki67', size=4, vmin=-0.05, vmax=0.05,\n",
    "                cmap=\"coolwarm\", ax=axs[1][1], show=False, figsize=(3,3), title=\"\", colorbar=False)\n",
    "\n",
    "\n",
    "# Fu\n",
    "\n",
    "scv.pl.scatter(fu_velocyto, layer=\"velocity\", color='TBX21', size=4, vmin=-0.05, vmax=0.05,\n",
    "                cmap=\"coolwarm\", ax=axs[2][0], show=False, figsize=(3,3), title=\"\", colorbar=False)\n",
    "\n",
    "scv.pl.scatter(fu_alevin, layer=\"velocity\", color='TBX21', size=4, vmin=-0.05, vmax=0.05,\n",
    "                cmap=\"coolwarm\", ax=axs[2][1], show=False, figsize=(3,3), title=\"\", colorbar=False)\n",
    "\n",
    "# add colorbar\n",
    "cmap = plt.get_cmap('coolwarm')\n",
    "norm = mpl.colors.Normalize(vmin=-0.05, vmax=0.05)\n",
    "sm = mpl.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "sm.set_array([])\n",
    "\n",
    "cbar_ax = fig.add_axes([0.4, 0.1, 0.2, 0.01])  # Position [left, bottom, width, height]\n",
    "fig.colorbar(sm, cax=cbar_ax, orientation='horizontal')\n",
    "\n",
    "# adjust and save\n",
    "plt.subplots_adjust(wspace=0.1, hspace=0.1)\n",
    "plt.savefig(\"figures/3_c.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6328fa1c-bfaf-48e8-b8dd-6ae5a5367418",
   "metadata": {},
   "source": [
    "## Table S1: R^2 values of HVGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c95560a-c9c9-491d-b502-1ecfd467d7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy datasets for modified velocity calculation\n",
    "stewart_velocyto_mod = stewart_velocyto.copy()\n",
    "stewart_alevin_mod = stewart_alevin[stewart_velocyto_mod.obs_names, stewart_velocyto_mod.var_names]\n",
    "\n",
    "mathew_velocyto_mod = mathew_velocyto.copy()\n",
    "mathew_alevin_mod = mathew_alevin[mathew_velocyto_mod.obs_names, mathew_velocyto_mod.var_names]\n",
    "\n",
    "fu_velocyto_mod = fu_velocyto.copy()\n",
    "fu_alevin_mod = fu_alevin[fu_velocyto_mod.obs_names, fu_velocyto_mod.var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8f72be7-213a-4c8d-955c-f6b78425fb35",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# calculate velocities with no threshold for R^2\n",
    "scv.tl.velocity(stewart_velocyto_mod, mode='stochastic', filter_genes=False, min_r2=-1)\n",
    "scv.tl.velocity(stewart_alevin_mod, mode='stochastic', filter_genes=False, min_r2=-1)\n",
    "\n",
    "scv.tl.velocity(mathew_velocyto_mod, mode='stochastic', filter_genes=False, min_r2=-1)\n",
    "scv.tl.velocity(mathew_alevin_mod, mode='stochastic', filter_genes=False, min_r2=-1)\n",
    "\n",
    "scv.tl.velocity(fu_velocyto_mod, mode='stochastic', filter_genes=False, min_r2=-1)\n",
    "scv.tl.velocity(fu_alevin_mod, mode='stochastic', filter_genes=False, min_r2=-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4af1d7af-9701-4019-bee6-d927e43fde5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get number of highly variable genes with R^2 > 0.01\n",
    "\n",
    "# get dataframe\n",
    "tab1 = pd.DataFrame({\"Dataset\": [\"Stewart et al.\", \"Mathew et al.\", \"Fu et al.\"],\n",
    "                     \"Velocyto\": [sum([stewart_velocyto_mod[:,stewart_velocyto_mod.var[\"highly_variable\"] == True].var[\"velocity_r2\"] > 0.01]).sum(),\n",
    "                                  sum([mathew_velocyto_mod[:,mathew_velocyto_mod.var[\"highly_variable\"] == True].var[\"velocity_r2\"] > 0.01]).sum(),\n",
    "                                  sum([fu_velocyto_mod[:,fu_velocyto_mod.var[\"highly_variable\"] == True].var[\"velocity_r2\"] > 0.01]).sum()], \n",
    "                     \"Alevin-fry\": [sum([stewart_alevin_mod[:,stewart_alevin_mod.var[\"highly_variable\"] == True].var[\"velocity_r2\"] > 0.01]).sum(),\n",
    "                                    sum([mathew_alevin_mod[:,mathew_alevin_mod.var[\"highly_variable\"] == True].var[\"velocity_r2\"] > 0.01]).sum(),\n",
    "                                    sum([fu_alevin_mod[:,fu_alevin_mod.var[\"highly_variable\"] == True].var[\"velocity_r2\"] > 0.01]).sum()]})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "993284bc-5a6c-4e51-b6f6-dbd37bca1bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tab1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce283d04-c650-4008-a049-a0035fcb3072",
   "metadata": {},
   "source": [
    "## Figure 4"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a291a5f-4cd7-4c6e-a7e1-87debb09803f",
   "metadata": {},
   "source": [
    "### Figure 4a: cosine similarity of velocities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea8bc01e-a94d-4cbc-80c9-171f796664a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get cosine similarities of velocities per cell\n",
    "stewart_cosine = []\n",
    "mathew_cosine = []\n",
    "fu_cosine = []\n",
    "\n",
    "for i in range(stewart_velocyto.n_obs):\n",
    "    stewart_cosine.append(-(cosine(stewart_velocyto.layers[\"velocity\"][i,:], stewart_alevin.layers[\"velocity\"][i,:])-1))\n",
    "\n",
    "for i in range(mathew_velocyto.n_obs):\n",
    "    mathew_cosine.append(-(cosine(mathew_velocyto.layers[\"velocity\"][i,:], mathew_alevin.layers[\"velocity\"][i,:])-1))\n",
    "\n",
    "for i in range(fu_velocyto.n_obs):\n",
    "    fu_cosine.append(-(cosine(fu_velocyto.layers[\"velocity\"][i,:], fu_alevin.layers[\"velocity\"][i,:])-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c0dd747-35e2-4d0b-ac0e-e1958a101fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for violinplot\n",
    "fig4a = pd.DataFrame({\"Cosine similarity\": np.concatenate([stewart_cosine, mathew_cosine, fu_cosine]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_obs,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_obs,\n",
    "                                                 [\"Fu et al.\"] * fu_velocyto.n_obs])})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7656be06-0c94-49fc-9401-e8a5d60b72b2",
   "metadata": {},
   "source": [
    "### Figure 4b: Pearson correlation of velocities per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3cb6bc6-47db-400f-b565-f1aee56bd2e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get pearson correlation for each gene\n",
    "stewart_corr = []\n",
    "mathew_corr = []\n",
    "fu_corr = []\n",
    "\n",
    "for gene in stewart_velocyto.var_names:\n",
    "    stewart_corr.append(pearsonr(stewart_velocyto[:,gene].layers[\"velocity\"].flatten(), stewart_alevin[:,gene].layers[\"velocity\"].flatten())[0])\n",
    "\n",
    "for gene in mathew_velocyto.var_names:\n",
    "    mathew_corr.append(pearsonr(mathew_velocyto[:,gene].layers[\"velocity\"].flatten(), mathew_alevin[:,gene].layers[\"velocity\"].flatten())[0])\n",
    "\n",
    "for gene in fu_velocyto.var_names:\n",
    "    fu_corr.append(pearsonr(fu_velocyto[:,gene].layers[\"velocity\"].flatten(), fu_alevin[:,gene].layers[\"velocity\"].flatten())[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8def750-4722-4e05-863e-31261ec200f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for violinplot\n",
    "fig4b = pd.DataFrame({\"Pearson correlation\": np.concatenate([stewart_corr, mathew_corr, fu_corr]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_vars,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_vars,\n",
    "                                                 [\"Fu et al.\"] * fu_velocyto.n_vars])})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7d05b0c-132f-4f8f-a1d6-fd247a79e519",
   "metadata": {},
   "source": [
    "### Figure 4c: Wasserstein distance of of transition probability distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24895c34-bc2e-4b43-94b7-c321963b4aa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get transition matrices\n",
    "stewart_velo_transition = scv.utils.get_transition_matrix(stewart_velocyto)\n",
    "stewart_alevin_transition = scv.utils.get_transition_matrix(stewart_alevin)\n",
    "mathew_velo_transition= scv.utils.get_transition_matrix(mathew_velocyto)\n",
    "mathew_alevin_transition= scv.utils.get_transition_matrix(mathew_alevin)\n",
    "fu_velo_transition = scv.utils.get_transition_matrix(fu_velocyto)\n",
    "fu_alevin_transition = scv.utils.get_transition_matrix(fu_alevin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4aa6a7a6-4470-4c8b-aff5-7ce790804d45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get Wasserstein distance\n",
    "\n",
    "stewart_wd = []\n",
    "mathew_wd = []\n",
    "fu_wd = []\n",
    "\n",
    "stewart_values = np.arange(stewart_velocyto.n_obs)\n",
    "mathew_values = np.arange(mathew_velocyto.n_obs)\n",
    "fu_values = np.arange(fu_velocyto.n_obs)\n",
    "\n",
    "\n",
    "# see: https://stackoverflow.com/questions/76049158/wasserstein-distance-in-scipy-definition-of-support\n",
    "for i in range(stewart_velocyto.n_obs):\n",
    "    stewart_wd.append(wasserstein_distance(stewart_values, stewart_values, stewart_velo_transition[i,:].toarray()[0], stewart_alevin_transition[i,:].toarray()[0]))\n",
    "\n",
    "for i in range(mathew_velocyto.n_obs):\n",
    "    mathew_wd.append(wasserstein_distance(mathew_values, mathew_values, mathew_velo_transition[i,:].toarray()[0], mathew_alevin_transition[i,:].toarray()[0]))\n",
    "\n",
    "for i in range(fu_velocyto.n_obs):\n",
    "    fu_wd.append(wasserstein_distance(fu_values, fu_values, fu_velo_transition[i,:].toarray()[0], fu_alevin_transition[i,:].toarray()[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f94678c-b163-4bbd-a704-91f33d0e4b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for violinplot\n",
    "fig4c = pd.DataFrame({\"Wasserstein distance\": np.concatenate([stewart_wd, mathew_wd, fu_wd]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_obs,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_obs,\n",
    "                                                 [\"Fu et al.\"] * fu_velocyto.n_obs])})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8be2faab-be05-49f6-a937-940b7f605a19",
   "metadata": {},
   "source": [
    "### Figure 4 combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c938f8b9-2aaf-4c85-b947-7123ab45fabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 4\n",
    "\n",
    "# set up figure\n",
    "fig, axs = plt.subplots(1, 3, figsize=(11,2.3))\n",
    "\n",
    "# a\n",
    "plt.subplot(1,3,1)\n",
    "sns.boxplot(x=\"Dataset\", y=\"Cosine similarity\", data=fig4a, \n",
    "               gap=0.6, showfliers=False, palette=[\"#548235\"])\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "# b\n",
    "plt.subplot(1,3,2)\n",
    "sns.boxplot(x=\"Dataset\", y=\"Pearson correlation\", data=fig4b, \n",
    "               gap=0.6, showfliers=False, palette=[\"#548235\"])\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "# c\n",
    "plt.subplot(1,3,3)\n",
    "sns.boxplot(x=\"Dataset\", y=\"Wasserstein distance\", data=fig4c, \n",
    "               gap=0.6, showfliers=False, palette=[\"#548235\"])\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "sns.despine()\n",
    "plt.subplots_adjust(top=0.8, right=0.84, wspace=0.8)\n",
    "plt.legend(bbox_to_anchor=(1.7,0.6))\n",
    "\n",
    "plt.savefig(\"figures/4.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feafd887-03b3-40c7-8c1d-ce42aa8826c1",
   "metadata": {},
   "source": [
    "## Supplement 6"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ff0b86-edb4-4127-8522-6f0af9b3f8e7",
   "metadata": {},
   "source": [
    "### Supplement 6a: number of genes with changed velocity sign per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d629292-70e4-41bc-812a-197a497633c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# count number of sign difference per cell\n",
    "\n",
    "stewart_signs, mathew_signs, fu_signs = [], [], []\n",
    "\n",
    "for i in range(stewart_velocyto.n_obs):\n",
    "    stewart_signs.append(np.sum(np.sign(stewart_velocyto.layers[\"velocity\"][i,:]) != np.sign(stewart_alevin.layers[\"velocity\"][i,:])))\n",
    "\n",
    "for i in range(mathew_velocyto.n_obs):\n",
    "    mathew_signs.append(np.sum(np.sign(mathew_velocyto.layers[\"velocity\"][i,:]) != np.sign(mathew_alevin.layers[\"velocity\"][i,:])))\n",
    "\n",
    "for i in range(fu_velocyto.n_obs):\n",
    "    fu_signs.append(np.sum(np.sign(fu_velocyto.layers[\"velocity\"][i,:]) != np.sign(fu_alevin.layers[\"velocity\"][i,:])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c40581e-2b19-41ee-93fa-2031c142edce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get percentage\n",
    "stewart_sign_perc = [x  / stewart_velocyto.n_vars for x in stewart_signs]\n",
    "mathew_sign_perc = [ x  / mathew_velocyto.n_vars for x in mathew_signs]\n",
    "fu_sign_perc = [x / fu_velocyto.n_vars for x in fu_signs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "192bf068-6103-4d99-9240-905b6d5694b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe for violinplot\n",
    "sup6a = pd.DataFrame({\"Proportion\": np.concatenate([stewart_sign_perc, mathew_sign_perc, fu_sign_perc]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_obs,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_obs,\n",
    "                                                 [\"Fu et al.\"] * fu_velocyto.n_obs])})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fb31517-f95f-438d-a37a-1e449c83d657",
   "metadata": {},
   "source": [
    "### Supplement 6b: similarities between neighbors as reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1580b25-3bfa-439f-8bed-a9604e6bc9ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cosine similarity of velocities per cell: compare each cell to nearest neighbor\n",
    "\n",
    "# Stewart:\n",
    "cosine_ref_stewart = []\n",
    "\n",
    "for i in range(stewart_velocyto.n_obs):  \n",
    "    # get nearest neighbor of each cell\n",
    "    array = stewart_velocyto.obsp[\"distances\"][i].toarray()\n",
    "    neighbor = np.argmin(np.where(array > 0, array, np.inf))\n",
    "\n",
    "    # get cosine similarity \n",
    "    cosine_ref_stewart.append(-(cosine(stewart_velocyto.layers[\"velocity\"][i,:], stewart_velocyto.layers[\"velocity\"][neighbor,:])-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08b4e926-5aef-474b-850e-e12271adc2e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mathew:\n",
    "cosine_ref_mathew = []\n",
    "\n",
    "for i in range(mathew_velocyto.n_obs):\n",
    "   \n",
    "    # get nearest neighbor of each cell\n",
    "    array = mathew_velocyto.obsp[\"distances\"][i].toarray()\n",
    "    neighbor = np.argmin(np.where(array > 0, array, np.inf))\n",
    "\n",
    "    # get cosine similarity \n",
    "    cosine_ref_mathew.append(-(cosine(mathew_velocyto.layers[\"velocity\"][i,:], mathew_velocyto.layers[\"velocity\"][neighbor,:])-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "139bcfc1-5da7-4db0-a3ee-d52f9a96f025",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fu:\n",
    "cosine_ref_fu = []\n",
    "\n",
    "# reduce genes to HVGs\n",
    "fu_hvgs = fu_velocyto.var[\"highly_variable\"].values\n",
    "fu_velocities = fu_velocyto.layers[\"velocity\"][:,fu_hvgs]\n",
    "\n",
    "for i in range(fu_velocyto.n_obs):\n",
    "   \n",
    "    # get nearest neighbor of each cell\n",
    "    array = fu_velocyto.obsp[\"distances\"][i].toarray()\n",
    "    neighbor = np.argmin(np.where(array > 0, array, np.inf))\n",
    "\n",
    "    # get cosine similarity \n",
    "    cosine_ref_fu.append(-(cosine(fu_velocities[i], fu_velocities[neighbor])-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03e651e5-b892-46c9-9f2c-df561f1e0c51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dataframe\n",
    "sup6b = pd.DataFrame({\"Cosine similarity\": np.concatenate([cosine_ref_stewart, cosine_ref_mathew, cosine_ref_fu]),\n",
    "                      \"Dataset\": np.concatenate([[\"Stewart et al.\"] * stewart_velocyto.n_obs,\n",
    "                                                 [\"Mathew et al.\"] * mathew_velocyto.n_obs,\n",
    "                                                 [\"Fu et al.\"] * fu_velocyto.n_obs]),\n",
    "                      \"Method\": np.concatenate([[\"Velocyto\"] * stewart_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * mathew_velocyto.n_obs,\n",
    "                                                [\"Velocyto\"] * fu_velocyto.n_obs])})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "887ff48d-5242-4905-95f8-a3091eff959b",
   "metadata": {},
   "source": [
    "### Supplement 6 combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a5af957-0f40-48ea-b75b-709966746590",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot supplement 6\n",
    "\n",
    "# set up figure\n",
    "fig, axs = plt.subplots(1, 2, figsize=(8,2.5))\n",
    "\n",
    "# a\n",
    "plt.subplot(1,2,1)\n",
    "#plt.title(\"Opposite velocities per cell\")\n",
    "sns.boxplot(x=\"Dataset\", y=\"Proportion\", data=sup6a, width=0.4,\n",
    "               gap=0.4, showfliers=False, palette=[\"#548235\"])\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "# b\n",
    "plt.subplot(1,2,2)\n",
    "#plt.title(\"Similarity of velocities of nearest neighbors\")\n",
    "sns.boxplot(x=\"Dataset\", y=\"Cosine similarity\", data=fig6a_ref, width=0.4,\n",
    "               gap=0.4, showfliers=False, palette=[\"#548235\"])\n",
    "plt.xticks(rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.xlabel(\"\")\n",
    "\n",
    "sns.despine()\n",
    "plt.subplots_adjust(wspace=0.8)\n",
    "\n",
    "plt.savefig(\"figures/supplement_6.png\", format=\"png\", bbox_inches=\"tight\", dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50c398f0-2c2f-4269-a2aa-5a19f5a87606",
   "metadata": {},
   "source": [
    "## Figure 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "156cdcf6-e4ec-474c-b08f-4a2658e2146c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stewart: plot velocities in UMAP on grid\n",
    "scv.pl.velocity_embedding_grid(stewart_velocyto_sto, basis='umap', color=\"cluster\", arrow_length=2, \n",
    "                               density=0.9, arrow_size=(10,18,8),\n",
    "                               legend_loc=\"right margin\", title=\"\", dpi=300, save=\"figures/5a.png\")\n",
    "\n",
    "scv.pl.velocity_embedding_grid(stewart_alevin_sto, basis='umap', color=\"cluster\", arrow_length=2,\n",
    "                               density=0.9, arrow_size=(10,18,8),\n",
    "                               legend_loc=\"right margin\", title=\"\", dpi=300, save=\"figures/5b.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07fc8f14-df6d-4c53-b15c-5d78c637e2e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stewart: PAGA\n",
    "scv.tl.paga(stewart_velocyto, groups='cluster')\n",
    "scv.tl.paga(stewart_alevin, groups='cluster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16dca583-0943-4cf3-bf19-cd09b7ad9ad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stewart: plot PAGA on UMAP\n",
    "scv.pl.paga(stewart_velocyto_sto, basis='umap', color='cluster', size=10, alpha=0.7, \n",
    "            min_edge_width=2, node_size_scale=1.5, title=\"\", dpi=300, save=\"figures/5c.png\")\n",
    "\n",
    "scv.pl.paga(stewart_alevin_sto, basis='umap', color='cluster', size=10, alpha=0.7, \n",
    "            min_edge_width=2, node_size_scale=1.5, title=\"\", dpi=300, save=\"figures/5d.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1bdd2365-ac78-4d97-bb06-a41a17d8900b",
   "metadata": {},
   "source": [
    "## Supplement 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf46e12-d0ca-469c-a80d-d07d2a02a9bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mathew: create colormap for UMAP\n",
    "\n",
    "# list of HEX colors\n",
    "hex_colors = [\"#CD665F\", \"#BF781C\", \"#AB871D\", \"#92941D\", \"#77A119\", \"#57AA17\", \"#64B15A\", \"#5CB488\",\n",
    "              \"#61AFB4\", \"#5AA6DC\", \"#5A91F8\", \"#7D7AFA\", \"#A760FA\", \"#C54FE4\", \"#CB4EC0\", \"#D3558E\"]\n",
    "\n",
    "# convert HEX colors to RGB\n",
    "rgb_colors = [(int(color[1:3], 16)/255, int(color[3:5], 16)/255, int(color[5:7], 16)/255) for color in hex_colors]\n",
    "\n",
    "# create colormap\n",
    "cmap = ListedColormap(rgb_colors)\n",
    "\n",
    "# create dictionary to assign colors to clusters\n",
    "clusters = [\"C1\", \"C2\", \"C3\", \"C4\", \"C5\", \"C6\", \"C7\", \"C8\", \"C9\", \"C10\", \"C11\", \"C12\", \"C13\", \"C14\", \"C15\", \"C16\"]\n",
    "umap_colors = {}\n",
    "i = 0\n",
    "for cluster in clusters:\n",
    "    umap_colors[cluster] = cmap.colors[i]\n",
    "    i+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec712481-6992-4484-8a8e-18c30594d7b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mathew: order cluster labels\n",
    "mathew_velocyto.obs['cluster'] = pd.Categorical(mathew_velocyto.obs['cluster'], categories=clusters, ordered=True)\n",
    "\n",
    "mathew_alevin.obs['cluster'] = pd.Categorical(mathew_alevin.obs['cluster'], categories=clusters, ordered=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42fa52b6-d324-4785-aaa4-4c48d2368bde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mathew: plot velocities in UMAP on grid\n",
    "scv.pl.velocity_embedding_grid(mathew_velocyto_sto, basis='umap', color=\"cluster\", arrow_length=2, palette=umap_colors,\n",
    "                               density=0.9, arrow_size=(10,18,8),\n",
    "                               legend_loc=\"right margin\", title=\"\", dpi=300, save=\"figures/supplement_7a.png\")\n",
    "\n",
    "scv.pl.velocity_embedding_grid(mathew_alevin_sto, basis='umap', color=\"cluster\", arrow_length=2, palette=umap_colors,\n",
    "                               density=0.9, arrow_size=(10,18,8),\n",
    "                               legend_loc=\"right margin\", title=\"\", dpi=300, save=\"figures/supplement_7b.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86ed6a27-0ffe-4332-b046-e032b1f3e7a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# choose cell from cluster 3 as root cell --> as described in code from Mathew et al.\n",
    "root = [i for i, x in enumerate(mathew_velocyto_sto.obs['cluster']) if x == 'C3'][0] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0314b2d-2b94-4871-9647-a99c91de9754",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mathew: PAGA\n",
    "scv.tl.paga(mathew_velocyto_sto, groups='cluster', root_key=root)\n",
    "scv.tl.paga(mathew_alevin_sto, groups='cluster', root_key=root)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "168d843e-15c2-4709-9e66-8b4517bcf769",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mathew: plot PAGA on UMAP\n",
    "scv.pl.paga(mathew_velocyto, basis='umap', color='cluster', size=5, alpha=0.5, palette=umap_colors,\n",
    "            min_edge_width=2, node_size_scale=1.5, title=\"\", dpi=300, save=\"figures/supplement_7c.png\")\n",
    "\n",
    "scv.pl.paga(mathew_alevin, basis='umap', color='cluster', size=5, alpha=0.5, palette=umap_colors,\n",
    "            min_edge_width=2, node_size_scale=1.5, title=\"\", dpi=300, save=\"figures/supplement_7d.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c19e4133-854c-444e-bcbe-b0ee94066145",
   "metadata": {},
   "source": [
    "## Supplement 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33dc848b-897c-4495-ba88-bf3e0b0411e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fu: plot velocities in UMAP on grid\n",
    "scv.pl.velocity_embedding_grid(fu_velocyto_sto, basis='umap', color=\"cluster\", arrow_length=2,\n",
    "                               density=0.9, arrow_size=(10,18,8),\n",
    "                               legend_loc=\"right margin\", title=\"\", dpi=300, save=\"figures/supplement_8a.png\")\n",
    "\n",
    "scv.pl.velocity_embedding_grid(fu_alevin_sto, basis='umap', color=\"cluster\", arrow_length=2,\n",
    "                               density=0.9, arrow_size=(10,18,8),\n",
    "                               legend_loc=\"right margin\", title=\"\", dpi=300, save=\"figures/supplement_8b.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "767e3f4a-4175-4241-9c04-a851d17d9457",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fu: convert cluster labels to strings\n",
    "fu_velocyto.obs[\"cluster\"] = fu_velocyto.obs[\"cluster\"].astype(str)\n",
    "fu_alevin.obs[\"cluster\"] = fu_alevin.obs[\"cluster\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dd4cf33-befe-4dc2-b2fe-1715f4248c5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fu: PAGA\n",
    "scv.tl.paga(fu_velocyto, groups='cluster')\n",
    "scv.tl.paga(fu_alevin, groups='cluster')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db18280-1b7f-4316-a285-0a63d3c38380",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fu: plot PAGA on UMAP\n",
    "scv.pl.paga(fu_velocyto, basis='umap', color='cluster', size=5, alpha=0.5,\n",
    "            min_edge_width=2, node_size_scale=1.5, title=\"\", dpi=300, save=\"figures/supplement_8c.png\")\n",
    "\n",
    "scv.pl.paga(fu_alevin, basis='umap', color='cluster', size=5, alpha=0.5,\n",
    "            min_edge_width=2, node_size_scale=1.5, title=\"\", dpi=300, save=\"figures/supplement_8d.png\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env1",
   "language": "python",
   "name": "env1"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
